src/DelayBucket.cc

```diff
@@ -25,9 +27,19 @@ DelayBucket::stats(StoreEntry *entry)const
 void
 DelayBucket::update(DelaySpec const &rate, int incr)
 {
-    if (rate.restore_bps != -1 &&
-            (level() += rate.restore_bps * incr) > rate.max_bytes)
-        level() = rate.max_bytes;
+    if (rate.restore_bps == -1)
+        return;
+
+    const auto oldLevel = level();
+    const auto delta = IntegralProduct(oldLevel, rate.restore_bps, incr);
+    if (delta.has_value()) {
+        const auto newLevel = IncreaseSum(oldLevel, delta.value());
+        if (newLevel.has_value()) {
+            level() = newLevel.value();
+            return;
+        }
+    }
+    level() = rate.max_bytes;
```

Comment:
Should we protect this assignment from overflowing (if the types change)? Or are we secretly relying on these two types to be identical?